Enhance Fast Grid Performance and Functionality

deepak-lenka opened the pull request:
1-iPhone Safari Memory Limit Handling
Disabled multithreading for iPhone Safari by detecting the user agent and avoiding SharedArrayBuffer usage to accommodate low memory limits.
2-Synthetic Event-Loop Expansion
Enhanced the event loop in view-worker.ts to include scrolling and rendering cell contents, ensuring smooth and efficient updates.
3-Scrollbar Navigation Improvement
Updated scrollbar navigation to be relative to the scroll track by adjusting scroll deltas based on the zoom level, providing a consistent scrolling experience.
4-Column Sorting and Filtering
Expanded sorting and filtering capabilities to support all columns, enabling dynamic configurations for multi-column operations.

deepak-lenka committed:
Expand Synthetic Event-Loop to include scrolling and rendering cell contents

deepak-lenka committed:
Disable multithreading for iPhone Safari to accommodate low memory limits

deepak-lenka committed:
Enable sorting and filtering for all columns

deepak-lenka committed:
Update README.md

1-iPhone Safari Memory Limit: Implemented logic to disable multithreading for iPhone Safari by checking the user agent and conditionally avoiding SharedArrayBuffer usage.
2-Expand Synthetic Event-Loop: Enhanced the event loop to include scrolling and rendering cell contents in view-worker.ts.
3-Scrollbar Navigation: Updated scrollbar navigation to be relative to the scroll track by adjusting scroll deltas based on the zoom level.
4-Sort/Filter All Columns: Expanded sorting and filtering capabilities to support all columns by updating the computeView function and View interface in view-worker.ts.

deepak-lenka committed:
Fix import conflict and remove unused variables

gabrielpetersson commented on src/row-manager/view-worker.ts:
```diff
@@ -263,21 +248,38 @@ const handleEvent = async (event: Message) => {
       cache.sort = null;
       return;
     }
+    case "scroll": {
```
this seems to do nothing?

gabrielpetersson replied:
let's remove this if it does nothing!

deepak-lenka replied:
Remove it

gabrielpetersson replied:
can we remove all the scroll related code in workers? so there is no dead code

deepak-lenka replied:
done remove all the scroll related code in workers

gabrielpetersson commented on src/scrollbar.ts:
```diff
@@ -131,6 +131,12 @@ export class Scrollbar {
     let deltaX = e.deltaX;
+
+    // Adjust deltas based on zoom level
+    const zoomLevel = window.innerWidth / document.documentElement.clientWidth;
```
nice wanted to add this for so long

deepak-lenka replied:
Thanks

gabrielpetersson commented on src/scrollbar.ts:
```diff
@@ -146,9 +152,6 @@ export class Scrollbar {
     this.isScrolling = true;
-    // NOTE(gab): makes sure scroll events are only triggered at most
-    // once every frame. useses transient scrolling to keep track of
-    // intermediate scroll offsets
```
why is this removed?

gabrielpetersson reviewed (commented):
awesome this is great! a couple qs

deepak-lenka committed:
Address PR feedback: restore scroll documentation, remove unused code

- Restore important documentation about transient scrolling behavior
- Remove unused updateView and renderCells functions from worker
- Keep zoom level adjustments for consistent scroll behavior

deepak-lenka committed:
Fix TypeScript errors and improve type safety

- Replace ViewConfig with View type from row-manager
- Add proper type annotations for map callbacks
- Fix string type assertions for filter values
- Remove duplicate interface definition

deepak-lenka committed:
refactor: remove unused scroll-related code from view-worker

gabriel-openai commented:
https://fast-grid-lz2z2v2cn-gabrielpeterssons-projects.vercel.app/ hmm this is the last deployment and seems like if i zoom in with cmd +, the scroll bar will still be too fast?

deepak-lenka commented:
I've checked and the scrollbar is still too fast when zooming in with cmd+

deepak-lenka commented:
If we make these changes scrolling feel much more natural when using cmd+ to zoom in! 

// After:
const zoomLevel = window.innerWidth / document.documentElement.clientWidth;
const zoomAdjustment = zoomLevel * zoomLevel;
const minDelta = 0.5;

deltaY /= zoomAdjustment;
deltaX /= zoomAdjustment;

if (deltaY !== 0 && Math.abs(deltaY) < minDelta) {
  deltaY = minDelta * Math.sign(deltaY);
}
if (deltaX !== 0 && Math.abs(deltaX) < minDelta) {
  deltaX = minDelta * Math.sign(deltaX);
}

gabrielpetersson commented:
actually yeah we might be talking about different thing -  im talking about dragging the scrollbar

your deployment: https://fast-grid-lz2z2v2cn-gabrielpeterssons-projects.vercel.app/ 
old deployment: https://fast-grid.vercel.app/

all zoom levels still have the same problem that it scrolls to fast / slow when dragging the scroll bar thumb around

deepak-lenka commented:
okay by adding zoom level adjustment to both of these methods, we can  fixed the exact issue the scrollbar thumb moving too fast or too slow when dragging at different zoom levels. 

// Added to onThumbDragY and onThumbDragX
const zoomLevel = window.innerWidth / document.documentElement.clientWidth;

// Changed calculation to include zoom level adjustment
((e.movementY / zoomLevel) / this.grid.viewportHeight) * state.tableHeight;

let me know ?
